bin/fa5-upgrade.sh

```diff
@@ -1,45 +1,136 @@
+     return 1;
+  fi
+
+  if [ -z "$tmp_dir" ]; then
+    echo "[FAIL] Generated temporary directory name is empty";
+    return 1;
+  fi
+
+  TEMP_DIR_PATH="$tmp_dir"
+}
+
+download_and_unpack_fontawesome_pro()
+{
+  local archive_file_name="$(npm pack @fortawesome/fontawesome-pro --silent)"
+  retval=$?
+  if [ "$retval" != 0 ]; then
+     echo "[FAIL] Can't download [$archive_file_name] archive";
+     return 1;
+  fi
+
+  tar -xzf "$archive_file_name"
+  retval=$?
+  if [ "$retval" != 0 ]; then
+     echo "[FAIL] Can't unpack [$archive_file_name] archive";
+     return 1;
+  fi
+
+  local font_dir_name="package"
+  if [ ! -d "$font_dir_name" ]; then
+    echo "[FAIL] Archive doesn't contain [$font_dir_name] required directory";
+    return 1;
+  fi
+
+  FONTAWESOME_PRO_DIR_NAME="$font_dir_name"
+}
+
+copy_ttf_fonts_to_dest_dir()
+{
+  mkdir -p "$DEST_DIR_PATH"
+  retval=$?
+  if [ "$retval" != 0 ]; then
+     echo "[FAIL] Can't create [$DEST_DIR_PATH] directory";
+     return 1;
+  fi
+
+  local font_dir_path="$TEMP_DIR_PATH/$FONTAWESOME_PRO_DIR_NAME/webfonts"
+
+  cp "$font_dir_path/fa-brands-400.ttf" "$DEST_DIR_PATH/FontAwesome5_Pro_Brands.ttf" &&
+  cp "$font_dir_path/fa-light-300.ttf" "$DEST_DIR_PATH/FontAwesome5_Pro_Light.ttf" &&
+  cp "$font_dir_path/fa-regular-400.ttf" "$DEST_DIR_PATH/FontAwesome5_Pro_Regular.ttf" &&
+  cp "$font_dir_path/fa-solid-900.ttf" "$DEST_DIR_PATH/FontAwesome5_Pro_Solid.ttf"
+
+  retval=$?
+  if [ "$retval" != 0 ]; then
+     echo "[FAIL] Can't copy ttf fonts to [$DEST_DIR_PATH] directory";
+     return 1;
+  fi
+}
+
+modify_package_json()
+{
+  /usr/bin/env node "./node_modules/$PROJECT_NAME/bin/add-font-assets.js"
+}
+
+react_native_link_project()
+{
+  react-native link "$PROJECT_NAME"
+}
+
+if setup_npm_config; then
+  echo "[SUCCESS] Set up npm config";
+else
+  exit 1;
 fi
 
-echo "Please enter your FontAwesome5 npm token:"
-read fa5_token
-npm config set "//npm.fontawesome.com/:_authToken" ${fa5_token}
-
-echo "Creating temporary folder"
-
-TEMP_DIR=$(mktemp -d -t rnvi)
-echo "Created folder $TEMP_DIR"
-pushd ${TEMP_DIR}
-
-echo "Downloading FontAwesome5 Pro"
-
-ARCHIVE=$(npm pack @fortawesome/fontawesome-pro --silent)
-tar -xzf ${ARCHIVE}
-mv package pro
-
-popd
-
-echo "Copying font files"
-
-mkdir -p assets/fonts
-
-cp ${TEMP_DIR}/pro/webfonts/fa-brands-400.ttf assets/fonts/FontAwesome5_Pro_Brands.ttf
-cp ${TEMP_DIR}/pro/webfonts/fa-light-300.ttf assets/fonts/FontAwesome5_Pro_Light.ttf
-cp ${TEMP_DIR}/pro/webfonts/fa-regular-400.ttf assets/fonts/FontAwesome5_Pro_Regular.ttf
-cp ${TEMP_DIR}/pro/webfonts/fa-solid-900.ttf assets/fonts/FontAwesome5_Pro_Solid.ttf
+if create_tmp_directory; then
+  echo "[SUCCESS] Temporary directory [$TEMP_DIR_PATH] was created";
+else
+  exit 1;
+fi
 
-echo "Modifying package.json"
+cd "$TEMP_DIR_PATH"
+if download_and_unpack_fontawesome_pro; then
+  echo "[SUCCESS] $FONT_NAME was unpacked to [$TEMP_DIR_PATH/$FONTAWESOME_PRO_DIR_NAME] directory";
+else
+  exit1 1;
```

Comment:
```suggestion
  exit 1;
```

Probably?

Reply:
I thought that exist from function is unexpected behavior as in other programming languages.
So, I decided to exit from script only at top level, while using functions as small reusable chunks.

You might want not to exit from function, but handle it's status and do something different to recover from error. That's why I added returns and not exit.

-

Reply:
You might be right about the unspecified behaviour, I have not read the specification for bash or shell but it seems it's not specified.